### PR TITLE
[Hydro] Issue106 use rtci to reduce load in prog

### PR DESCRIPTION
**Pull request by fmessmer**
This PR is a major change in the `ur_driver` package.

It solves issue #106!
(The problem that caused #106 was a "Possible CPU-overload due to structure of user program")

This PR now uses the UR controller socket connections on port 30002 (10Hz) and port 30003 (125Hz) with the respective UniversalRobot protocol data structures (see `deserialize.py` and `deserializeRT.py`) in order to get rid of **all** the `socket_send_int` calls in `prog`.
- `/joint_states` and `/wrench` are still published at 125 Hz using the same types as before, i.e. `sensor_msgs/JointState` and `geometry_msgs/WrenchStamped`
- DigitalInput and DigitalOutput are published on `/io_states` at 10 Hz.
- Unfortunately, (UR-Script/Software-)Flags are not present in neither port 30002 nor 30003. However, as the Flags are only used in UR-Scripts but in `prog` they are **not** used, I decided to remove support for flags!

The PR thus restores a _usable_ version of the `ur_driver.py`

---

As already mentioned in #106, the IO-Support needs a significant revision in order to reflect e.g. analog_input/output_range/domain as can be seen in the MasterboardData structure of `deserialize.py`
(see #118)

### CODE REVIEW

**Comment by fmessmer** [comment]
Also this PR already includes the following open PRs
- #78 and #102 
- #103 
- #110 
- #111 
- #114 
- #115 
- #116 

Sorry, for this but it would have been a mess otherwise....

File: ur_driver/prog [hunk 1]
```diff
@@ -35,60 +38,91 @@ def driverProg():
...
-    #socket_open(HOSTNAME, 50002)
-    send_joint_state()
-    while True:
-      send_joint_state()
-      sync()
-    end
-    sync()
-  end
+  
+  ## For joint states the real-time communication interface connection is used
```
Comment (shaun-edwards) [inline ur_driver/prog]: Can we remove commented out code?  Is there a need to keep it?

File: ur_driver/prog [hunk 2]
```diff
@@ -141,7 +175,9 @@ def driverProg():
   socket_open(HOSTNAME, PORT)
   send_out("hello")
   
-  thread_state = run statePublisherThread()
+  #thread_state = run statePublisherThread()
```
Comment (shaun-edwards) [inline ur_driver/prog]: More commented out code.

File: ur_driver/prog [hunk 3]
```diff
@@ -267,7 +313,9 @@ def driverProg():
   end
   
   #sleep(1)
-  kill thread_state
+  #kill thread_state
```
Comment (shaun-edwards) [inline ur_driver/prog]: More commented out code.

**Comment by gavanderhoorn** [comment]
@ipa-fxm, @shaun-edwards: if the PR would squash / remove the commits related to the previous attempts to fix this, review would be easier and the PR less messy (ie: most of the commented lines would be gone).

File: ur_driver/src/ur_driver/deserializeRT.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
+import struct
```
Comment (shaun-edwards) [inline ur_driver/src/ur_driver/deserializeRT.py:1]: Does deserializeRT.py adhere to file naming conventions.  In C++, files are under_scored.

Comment (shaun-edwards) [inline ur_driver/src/ur_driver/deserializeRT.py:1]: File needs license header (others probably need it as well).

**Comment by gavanderhoorn** [comment]
Also: I don't think it's really necessary to include all changes from the other PRs? Added functionality is pretty orthogonal to those changes. The outstanding PRs that would cause merging conflicts (mostly mine) could probably be closed (they mostly fixup things which will be removed by the changes in this PR).

File: ur_driver/src/ur_driver/driver.py [hunk 5]
```diff
@@ -474,7 +511,7 @@ def set_tool_voltage(self, value):
         #print params
         with self.socket_lock:
             self.request.send(buf) 
-        time.sleep(IO_SLEEP_TIME+.5)
+        time.sleep(IO_SLEEP_TIME)
```
Comment (shaun-edwards) [inline ur_driver/src/ur_driver/driver.py:514]: Not sure if this line is part of this PR, but I will comment here.  This sleep time introduces unnecessary delays or sometimes is does not delay long enough.  We should consider a better (possibly handshake) driven approach (not a requirement for this PR, since IO is experimental).

File: ur_driver/src/ur_driver/io_interface.py [hunk 6]
```diff
@@ -10,49 +10,50 @@
...
+        set_io(FUN_SET_TOOL_VOLTAGE, volts, 0)
+    except rospy.ServiceException, e:
+        print "Service call failed: %s"%e
 
 def set_digital_out(pin, val):
-    set_io_val(FUN_SET_DIGITAL_OUT, pin, val)
-    if Digital_Out_States[pin] != val:
-        rospy.logwarn("DIGITAL OUT IS NOT A HAPPY CAMPER, its not setting a pin on the first try, it may help to increase IO_SLEEP_TIME in driver.py")
-        set_digital_out(pin, val)
+    try:
```
Comment (shaun-edwards) [inline ur_driver/src/ur_driver/io_interface.py:34]: Again, this might not be part of this PR.  The deleted section checks to see if the IO state is correctly set, but the new section only checks if the service call failed (these are different failure modes).

File: ur_driver/src/ur_driver/testRT_comm.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
```
Comment (shaun-edwards) [inline ur_driver/src/ur_driver/testRT_comm.py:1]: File naming convention?

Comment (shaun-edwards) [inline ur_driver/src/ur_driver/testRT_comm.py:1]: Needs license header.

Comment (shaun-edwards) [inline ur_driver/src/ur_driver/testRT_comm.py:1]: +1 for adding a test script!

File: ur_driver/src/ur_driver/test_comm.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
```
Comment (shaun-edwards) [inline ur_driver/src/ur_driver/test_comm.py:1]: License header.

File: ur_msgs/msg/MasterboardDataMsg.msg [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+int8 analog_input_range0
+int8 analog_input_range1
+float64 analog_input0
+float64 analog_input1
+int8 analog_output_domain0
+int8 analog_output_domain1
+float64 analog_output0
+float64 analog_output1
+float32 masterboard_temperature
+float32 robot_voltage_48V
```
Comment (shaun-edwards) [inline ur_msgs/msg/MasterboardDataMsg.msg:23]: Should this be `robot_voltage_48v` to match ros message naming conventions?

File: ur_msgs/msg/MasterboardDataMsg.msg [hunk 10]
```diff
@@ -0,0 +1,16 @@
+int16 digital_input_bits
```
Comment (shaun-edwards) [inline ur_msgs/msg/MasterboardDataMsg.msg:12]: Some of these message members aren't obvious.  Can this be better commented or provide a link to the relevant UR docs.

File: ur_msgs/msg/RobotStateRTMsg.msg [hunk 11]
```diff
@@ -0,0 +1,19 @@
+float64 time
```
Comment (shaun-edwards) [inline ur_msgs/msg/RobotStateRTMsg.msg:9]: Comments/documentation for this message would be helpful as well.

**Comment by shaun-edwards** [comment]
I agree that with @gavanderhoorn that multiple PRs are messy.  Sorry if my comments span multiple PRs.  Overall, things look good.  Thanks so much for the effort!

File: ur_driver/prog (hunk 1 above)
Comment (fmessmer) [inline ur_driver/prog]: I just kept it as a fall back and for testing.
If it is desired to be removed, I can remove it....

Update: removed!

File: ur_driver/prog (hunk 2 above)
Comment (fmessmer) [inline ur_driver/prog]: I just kept it as a fall back and for testing.
If it is desired to be removed, I can remove it....

Update: removed!

File: ur_driver/prog (hunk 3 above)
Comment (fmessmer) [inline ur_driver/prog]: I just kept it as a fall back and for testing.
If it is desired to be removed, I can remove it....

Update: removed!

File: ur_driver/src/ur_driver/deserializeRT.py (hunk 4 above)
Comment (fmessmer) [inline ur_driver/src/ur_driver/deserializeRT.py:1]: license header should be added in a separate PR to not further mess up this PR.
see #119

File: ur_driver/src/ur_driver/driver.py (hunk 5 above)
Comment (fmessmer) [inline ur_driver/src/ur_driver/driver.py:514]: The IO-Interface definitely needs total revision (see #118)

File: ur_driver/src/ur_driver/io_interface.py (hunk 6 above)
Comment (fmessmer) [inline ur_driver/src/ur_driver/io_interface.py:34]: The IO-Interface definitely needs total revision (see #118)

File: ur_msgs/msg/MasterboardDataMsg.msg (hunk 9 above)
Comment (fmessmer) [inline ur_msgs/msg/MasterboardDataMsg.msg:23]: V is an SI unit marked with capital V....also this is the name of the variable in the UR structure.

**Comment by fmessmer** [comment]
I suggest the following:
In order to reduce this PR to only commits relevant to the driver restructuring, I would accept the following commits as they are included in this PR. I will then rebase to reduce the commits.
#102
#103
#110
#111
#114
#115
#116
The according issues should be closed after merge as well.

All other (remaining open) issues should then be adressed in fresh new PRs. 
#118
#119

**Comment by gavanderhoorn** [comment]
Finally had some time to work with this on our UR5: it works nicely. See also my comments on #106, the drop in controller cpu usage is very nice. IO states are indeed at 10Hz, but the real-time client interface just doesn't contain the data we need.

I've noticed no functional changes.

@ipa-fxm: if we can clean this up some more (see also @shaun-edwards comments), this would be a good candidate for merging.

Ideally, we'd get some more people to test this out and in more realistic use cases (I just planned some very simple motions).

**Comment by gavanderhoorn** [comment]
The deserialiser for the RT packets also needs to be able to handle different firmware versions. The UR wiki documents the differences between the various versions.

File: ur_msgs/msg/MasterboardDataMsg.msg (hunk 10 above)
Comment (fmessmer) [inline ur_msgs/msg/MasterboardDataMsg.msg:12]: documentation and link added

File: ur_msgs/msg/RobotStateRTMsg.msg (hunk 11 above)
Comment (fmessmer) [inline ur_msgs/msg/RobotStateRTMsg.msg:9]: documentation and link added

**Comment by fmessmer** [comment]
After several other PRs have been merged, I rebased to latest `hydro-devel`

I also considered most of the comments here.
If comments are not considered in this PR, new issues have been created to document the comments.

As tests by @gavanderhoorn confirmed that this PR is working and solving #106, this can now be merged!

Please note: As this now introduces a bigger change in the driver layout, I'll advise everyone to have a critical look at the current code in order to find problems/bugs that might have slipped through!

**Comment by abubeck** [comment]
I agree with @ipa-fxm and will merge this now.

**Comment by gavanderhoorn** [comment]
On 10-10-2014 11:51, Felix Messmer wrote:

> After several other PRs have been merged, I rebased to latest `hydro-devel`
> 
> I also considered most of the comments here.
> If comments are not considered in this PR, new issues have been created to document the comments.
> 
> As tests by @gavanderhoorn confirmed that this PR is working and solving #106, this can now be merged!
> 
> Please note: As this now introduces a bigger change in the driver layout, I'll advise everyone to have a critical look at the current code in order to find problems/bugs that might have slipped through!

I think we should wait a little, and let other users test your branch
before merging. That way we know a little bit more about the stability.

I think I also suggested this in an earlier comment on this PR.

**Comment by gavanderhoorn** [comment]
Also: by merging this PR, compatibility with controllers running anything but 1.8 is removed. imho such a regression should've been addressed before merging this (I know #123 exists, but that doesn't change this).

**Comment by gavanderhoorn** [comment]
I just also noticed that the `effort` field in `JointStates` is now always 0. People making use of this field reflecting actual joint torques will not be able to continue using this.
